src/Tools/ExternalAccess/Format/EditorConfigOptionsApplier.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Reflection;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.VisualStudio.CodingConventions;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Format
+{
+    internal class EditorConfigOptionsApplier
+    {
+        private readonly ImmutableArray<(IOption, OptionStorageLocation, MethodInfo)> _formattingOptionsWithStorage;
+
+        public EditorConfigOptionsApplier()
+        {
+            var optionTypes = new Type[]
+            {
+                typeof(Formatting.FormattingOptions),
+                typeof(CSharp.Formatting.CSharpFormattingOptions),
+                typeof(Editing.GenerationOptions),
+                typeof(CodeStyleOptions),
+                typeof(CSharp.CodeStyle.CSharpCodeStyleOptions),
+            };
+
+            _formattingOptionsWithStorage = GetOptionsWithStorageFromTypes(optionTypes);
+        }
+
+        /// <summary>
+        /// Apply .editorconfig settings to the <see cref="OptionSet" />.
+        /// </summary>
+        public OptionSet ApplyConventions(OptionSet optionSet, ICodingConventionsSnapshot codingConventions, string languageName)
+        {
+            foreach (var optionWithStorage in _formattingOptionsWithStorage)
+            {
+                if (TryGetConventionValue(optionWithStorage, codingConventions, out var value))
+                {
+                    var option = optionWithStorage.Item1;
+                    var optionKey = new OptionKey(option, option.IsPerLanguage ? languageName : null);
+                    optionSet = optionSet.WithChangedOption(optionKey, value);
+                }
+            }
+
+            return optionSet;
+        }
+
+        internal ImmutableArray<(IOption, OptionStorageLocation, MethodInfo)> GetOptionsWithStorageFromTypes(params Type[] formattingOptionTypes)
+        {
+            var optionType = typeof(IOption);
+
+            var optionsFromProperties = formattingOptionTypes
+                .SelectMany(t => t.GetProperties(BindingFlags.Public | BindingFlags.Static | BindingFlags.GetProperty))
+                .Where(p => optionType.IsAssignableFrom(p.PropertyType))
+                .Select(p => (IOption)p.GetValue(null));
```

Comment:
:-/

This feels so brittle.  We have a reflection dependency in our own project.  We could easily break this and never have a single clue.